File: go-cli/internal/service/note.go
// Package service contains business logic.
package service

import (
	"convex-notes-cli/internal/entity"
	"convex-notes-cli/internal/repository"
)

// NoteService provides note business logic.
type NoteService struct {
	repo repository.NoteRepository
}

// NewNoteService creates a new NoteService.
func NewNoteService(repo repository.NoteRepository) *NoteService {
	return &NoteService{repo: repo}
}

// List returns all notes.
func (s *NoteService) List() ([]entity.Note, error) {
	return s.repo.List()
}

// Get returns a note by ID.
func (s *NoteService) Get(id string) (*entity.Note, error) {
	return s.repo.Get(id)
}

// Create creates a new note.
func (s *NoteService) Create(title, content string) (string, error) {
	return s.repo.Create(entity.CreateNote{
		Title:   title,
		Content: content,
	})
}

// Update updates an existing note.
func (s *NoteService) Update(id, title, content string) error {
	return s.repo.Update(entity.UpdateNote{
		ID:      id,
		Title:   title,
		Content: content,
	})
}

// Delete removes a note.
func (s *NoteService) Delete(id string) error {
	return s.repo.Delete(id)
}

File: zig-cli/src/service/note.zig
//! Note service for business logic.

const std = @import("std");

const note = @import("../entity/note.zig");
const Note = note.Note;
const convex = @import("../repository/convex.zig");
const ConvexRepository = convex.ConvexRepository;

/// Note service wrapping repository.
pub const NoteService = struct {
    repo: *ConvexRepository,

    pub fn init(repo: *ConvexRepository) NoteService {
        return .{ .repo = repo };
    }

    pub fn list(self: *NoteService) ![]Note {
        return self.repo.list();
    }

    pub fn create(self: *NoteService, title: []const u8, content: []const u8) ![]const u8 {
        return self.repo.create(.{ .title = title, .content = content });
    }

    pub fn update(self: *NoteService, id: []const u8, title: []const u8, content: []const u8) !void {
        return self.repo.update(.{ .id = id, .title = title, .content = content });
    }

    pub fn delete(self: *NoteService, id: []const u8) !void {
        return self.repo.delete(id);
    }
};
